### DISCUSSION TITLE
How to delete a file from repo commit

### DISCUSSION THREAD

**Discussion by siddharthshubhampal** [community/community#50813] [Repositories]
Hi 

I have multiple repos in which someone has committed confidential file. How can I delete this file from that repos and all of its branches without effecting any cloned repo/user

**Comment by queenofcorgis** [comment]
Welcome to the GitHub Community, @siddharthshubhampal , we're happy you're here! You are more likely to get a useful response if you are posting your questionin the applicable category. I've gone ahead and moved it for you. Good luck!

**Reply by siddharthshubhampal** [reply]
@queenofcorgis Thanks for moving the question
will take care of selecting correct tag next time onwards

**Comment by airtower-luna** [comment]
Removing/changing a commit without affecting any clones of the repository isn't possible. Changing a commit changes all commits that were made after. Anyone who already has the commit with the file you want to remove will have to manually accept the changed history (or delete their repository and make a fresh clone, whichever is more convenient).

Otherwise, the general process is:

1. Rewrite the branch(es) including the commit that adds the confidential file so the file isn't added. There are multiple ways to do that, see the documentation links below.
2. Force-push the changed branch(es) to the GitHub repo. Mind that force-pushing discards any remote changes, double check what you're pushing is right or you might lose data. :warning: 
3. Contact GitHub support and ask them to purge the original commit from all caches.

See also:
* [Git Tools - Rewriting History](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) about changing Git history
* GitHub docs on [Removing sensitive data from a repository](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository)
